bite-size-notes/week-6-a-CTCS_SCP.md

```diff
@@ -0,0 +1,171 @@
+
+$$
+\dot{y}(t) \coloneqq 1^\top|g(t, x(t), u(t))|^2_+ + 1^\top h(t, x(t), u(t))^2
+$$
+
+ From $y$ dynamics definition, $\dot{y}(t) \geq 0\; \forall t \in [t_0, t_f]$; if boundary condition $y_0 = y_f$ is satisfied, then $\dot{y}(t) = 0 \;  \forall t \in [t_0, t_f]$, implying constraints $g(t, x(t), u(t)) \leq 0, h(t, x(t), u(t))=0$ are satisfied $\forall t \in [t_0, t_f]$. Therefore, the discretized problem reads
+
+$$\begin{array}{rl}
+\underset{x_i,p_i,t_f}{\text{minimize}} & L(t_{N+1}, x_{N+1}) \\
+\text{subject to} & x_{i+1} = x_{i} + \int_{t_{i}}^{t_{i+1}} f(\tau, x(\tau), \xi_i(\tau, p_i))\text{d}\tau  \quad i = 1\dots, N \\
+& y_{i+1} = y_{i} + \int_{t_{i}}^{t_{i+1}} [1^\top |g(\tau, x(\tau), \xi_i(\tau, p_i))|^2_+ + 1^\top h(\tau, x(\tau), \xi_i(\tau, p_i))^2]\text{d}\tau  \quad i = 1\dots, N \\
+& y_{i+1} - y_{i} \leq \varepsilon \quad i = 1\dots, N \\
+& t_1 = t_0, t_{N+1} = t_f \\
+& Q(t_1, x_1, t_{N+1}, x_{N+1}) = 0 \\
+\end{array}
+$$
+
+where the positive variable $\varepsilon$ allows to 1) grant Linear Independence Constraint Qualification (LICQ), making the problem tractable, and 2) leverage an exact penalty approach [2]. In addition, the given $\varepsilon$ is linked with the maximum pointwise violation of the considered path constraints, according to the following relation 
+
+$$
+\int_{t_k}^{t_{k+1}} |g^j(\tau, x(\tau), \xi_i(\tau, p_i)) + \delta_{g^j}(\varepsilon)|^2_+ \text{d}t \leq \varepsilon \Rightarrow g^j(\tau, x(\tau), \xi_i(\tau, p_i)) \leq 0, \; \forall t \in [t_k, t_{k+1}]
+$$
+
+where the monotonic increasing function $\delta_{g^j}(\varepsilon)$ can be estimated at a first iteration; the SCP approach can then be adopted, for decreasing $\varepsilon$, up to satisfaction of a desired tolerance $\texttt{tol}$: $\delta_{g^j}(\varepsilon) < \text{\texttt{tol}}$.
+
+### Inter-sample constraint satisfaction
+The developed framework is now applied to the same drone example as before, but now embedding the obstacle avoidance constraint in the CTCS formulation. Constraint is rported again for clarity here below.
+
+$$
+g^j(t, x(t), u(t)) = 1 - \left\| S^j(r(t) - r^j_{\text{obs}})\right\| \leq 0
+$$
+
+The NMPC framework now provides far better results with respect to the case outlined before, and nodes are correctly placed in such manner to correctly satisfy the assigned constraints.
+
+![CTCS_yesCS](figs/ctcs_animation.gif)
+*Retrieved with permission from [4]
+
+## Conclusion
+The present note has outlined one of the most recent advances in SCP-based algorithms, that allows Continuous-Time Constraint Satisfaction in spite of the DT formulation of the original optimal control problem. The developed framework has been applied to an obstacle avoidance drone example; additional ones are as well available in literature [2], such as rocket-landing scenarios and some others are yet to be proposed.
+
```

Comment:
Overall, the notes present a clear picture of the problem with the Direct approach of discretizing and solving the optimal control problem, and using the CTCS method to fix that. Moreover, the drone example with the graphical simulation also helped in understanding the application of the NMPC.

Reply:
Thanks for the insightful and meticulous review. Hope comments have been thoroughly addressed